---
title: "Fst-through-time Patterns"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Fst-through-time Patterns}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

This vignette shows an example of how you can use `onthefly` to fit a population genetics pattern to a simulation (simulated using `slimr`).

First we load the packages we need:

```{r setup}
library(onthefly)
library(slimr)
```

We are going to create a simple simulation of multiple subpopulations with migration, drift, and selection, with population fluctuations driven by an external factor (in this case 'virtual' rainfall). Once we have our simulation we will use it to generate a single example of dynamics in a classic population genetics measure, the fixation index (F~st~). We will then use `onthefly` to try and find the parameters that may have led to the pattern seen in that example. We have chosen this example because we know that F~st~ responds to many different evolutionary processes and so we expect to have similar patterns generated by multiple combinations of parameters. This specifically highlights the ability of `onthefly` to model multimodel posterior distributions for simulation parameter spaces. Okay, so here is our simulation code:

```{r sim_code}
slim_script(
  
  slim_block(initialize(), {
    
    defineConstant("alpha", slimr_template("mut_rate", 1e-6))
    
    initializeSLiMOptions(nucleotideBased=T);
    initializeAncestralNucleotides(randomNucleotides(slimr_template("genome_size", 3000)));
    initializeMutationTypeNuc("m1", 0.5, "n", 0.0, slimr_template("selection_strength", 0.1));
    
    #initializeMutationRate(slimr_template("mut_rate", 1e-6));
    #initializeMutationType("m1", 0.5, "n", 0, slimr_template("selection_strength", 0.1));
    initializeGenomicElementType("g1", m1, 1.0, matrix(c(0.0, alpha, 0.0, 0.0,
                                                         alpha, 0.0, 0.0, 0.0,
                                                         0.0, 0.0, 0.0, alpha,
                                                         0.0, 0.0, alpha, 0.0),
                                                       nrow = 4,
                                                       ncol = 4));
    initializeGenomicElement(g1, 0, ..genome_size.. - 1);
    initializeRecombinationRate(slimr_template("recomb_rate", 1e-8));
    initializeSex("A");
    
    defineConstant("rainfall", slimr_inline(rainfall, delay = TRUE));
    defineConstant("sample_these", slimr_inline(sample_these, delay = TRUE));
    defineConstant("samp_sizes", slimr_inline(samp_sizes, delay = TRUE));
    defineConstant("init_pop", slimr_inline(init_popsize, delay = TRUE));
    defineConstant("N_pop", size(init_pop));
    
  }),
  slim_block(1, {
    
    ## set populations to initial size
    for(pop in seqLen(N_pop)) {
      sim.addSubpop(pop + 1, asInteger(init_pop[pop]));  
    }

  }),
  
  slim_block(1, late(), {
    
    ## migration on or off flags for pops 1-N_pop (using tag)
    for(pop in seqLen(N_pop)) {
      sim.subpopulations[pop]%.%tag = 0;
    }
    
  }),
  
  slim_block(1, 500, late(), {
    
    ## update generation number
    gen = sim.generation 
    
    for(pop in seqLen(N_pop)) {
      sim.subpopulations[pop]%.%tag = 0;
    }
    
    ## set population size to observed levels
    for(pop in seqLen(N_pop)) {
      sim.subpopulations[pop]$setSubpopulationSize(asInteger(ceil(rainfall[gen - 1] * slimr_template("popsize_scaling", 100))));
    }
   
    ## increase migration when above rainfall threshold
    
    for(pop1 in seqLen(N_pop)) {
      for(pop2 in seqLen(N_pop)) {
        if(pop1 != pop2) {
          if(sim.subpopulations[pop1]$tag == 0 & rainfall[gen - 1] > slimr_template("rainfall_threshold", 10)) {
            sim.subpopulations[pop2]$setMigrationRates(pop1 + 1, slimr_template("migration_rate", 0));
          }
        }
      }
      sim.subpopulations[pop1]%.%tag = 1;
    }
    
    ## only run if the generation is in our sample_these list
    if(any(match(sample_these, sim.generation) >= 0)) {
      ## find the sample size that matches the matching "year" for our obs data
      ssizes = drop(samp_sizes[ , which(sample_these == sim.generation)]) 
      ## if number of individuals is less than sample size, just sample all
      ssizes = pmin(ssizes, sim.subpopulations.individualCount)
      ## sample individuals
      ind_sample = sapply(seqLen(N_pop), "sample(sim.subpopulations[applyValue].individuals, ssizes[applyValue]);")
      
      ## output individuals genomes
      slimr_output_nucleotides(both_genomes = TRUE, inds = ind_sample)
      slimr_output(ind_sample.subpopulation, "subpops")
      # slimr_output(ind_sample.genomes.output(), "pop_sample", do_every = 1);
      # slimr_output(ind_sample.genomes.individual.subpopulation, "subpops", do_every = 1)
    }
    
  }),
  
  slim_block(500, late(), {
    sim.simulationFinished()
  })
  
) -> pop_sim

pop_sim
```

So there is a fair amount happening in that simulation but let's just jump into running some with example parameters to get a sense of what dynamics we see. This simulations outputs binary nucleotide SNPs for a sample of the population, so once we've run our simulations we will need to calculate F~st~ on them (using an efficient method for fast computation, which will be important later). Let's test one simulation. To do this we will need to render the `slimr_script` we made above to fill in our templated variables and our inlined objects.

```{r render_script}

N_pop <- 5 ## number of subpops

## cartoon periodic rainfall
A <- 10 ## Amplitude
f <- 0.02 ## Frequency
t <- 1:500 ## number of generations
rainfall <- A + (A * sin(2 * pi * f * t))
plot(rainfall, type = "l")

## sample at random time-points -- let's say we have 50 time samples
sample_these <- sort(sample.int(500, 50))
plot(rainfall, type = "l")
abline(v = sample_these, col = "red")

## sample a constant number (50) of individuals per subpop per timepoint
samp_sizes <- matrix(50, nrow = N_pop, ncol = 500)

## initial population size
init_popsize <- rep(100, N_pop)

## render script with default parameters
pop_sim_1 <- slim_script_render(pop_sim)
pop_sim_1
```
Okay, let's see if that simulation runs:

```{r run_sim_test}

res <- slim_run(pop_sim_1)
dat <- slim_results_to_data(res)

```



